presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestThrottledAsyncQueue.java

```diff
@@ -0,0 +1,266 @@
+
+        // no throttling, enough elements in the queue
+        ListenableFuture<List<Integer>> future1 = queue.getBatchAsync(2);
+        assertTrue(future1.isDone());
+        assertEquals(getFutureValue(future1), ImmutableList.of(1, 2));
+        assertFalse(queue.isFinished());
+
+        // we can only dequeue one more element before being throttled
+        ListenableFuture<List<Integer>> future2 = queue.getBatchAsync(2);
+        assertFalse(future2.isDone());
+        assertEquals(getFutureValue(future2), ImmutableList.of(3, 4));
+        assertFalse(queue.isFinished());
+
+        // we are now throttled, this future will not be immediate
+        ListenableFuture<List<Integer>> future3 = queue.getBatchAsync(2);
+        assertFalse(future3.isDone());
+        assertEquals(getFutureValue(future3), ImmutableList.of(5, 6));
+        assertTrue(queue.isFinished());
+    }
+
+    @Test(timeOut = 10_000)
+    public void testThrottleEmptyQueue()
+            throws Exception
+    {
+        // Make sure that the de-queuing is throttled even if we have enough elements in the queue
+
+        ThrottledAsyncQueue<Integer> queue = new ThrottledAsyncQueue<>(2, 10, executor);
+        assertTrue(queue.offer(1).isDone());
+        assertTrue(queue.offer(2).isDone());
+
+        // no throttling, enough elements in the queue
+        ListenableFuture<List<Integer>> future1 = queue.getBatchAsync(2);
+        assertTrue(future1.isDone());
+        assertEquals(getFutureValue(future1), ImmutableList.of(1, 2));
+        assertFalse(queue.isFinished());
+
+        // we are now throttled and the queue is empty
+        ListenableFuture<List<Integer>> future2 = queue.getBatchAsync(2);
+        assertFalse(future2.isDone());
+
+        Thread.sleep(1000L); // wait one second, after which we should not be throttled any more
+
+        // no batch is ready at that point as the queue is still empty
+        assertFalse(future2.isDone());
+
+        assertTrue(queue.offer(3).isDone());
+        assertTrue(queue.offer(4).isDone());
+        queue.finish();
+
+        assertEquals(getFutureValue(future2), ImmutableList.of(3, 4));
+
+        assertTrue(queue.isFinished());
+    }
+
+    @Test(timeOut = 10_000)
+    public void testBorrowThrows()
+            throws Exception
+    {
+        // It doesn't matter the exact behavior when the caller-supplied function to borrow fails.
+        // However, it must not block pending futures.
+
+        AsyncQueue<Integer> queue = new ThrottledAsyncQueue<>(100, 4, executor);
+        queue.offer(1);
+        queue.offer(2);
+        queue.offer(3);
+        queue.offer(4);
+        queue.offer(5);
+
+        ListenableFuture<?> future1 = queue.offer(6);
+        assertFalse(future1.isDone());
+
+        Runnable runnable = () -> {
+            getFutureValue(queue.borrowBatchAsync(1, elements -> {
+                throw new RuntimeException("test fail");
+            }));
+        };
+
+        try {
```

Comment:
This can be simplified using AssertJ
```java
assertThatThrownBy(() -> executor.submit(runnable).get())
        .isInstanceOf(ExecutionException.class)
        .hasMessageContaining("test fail");

Reply:
done